CMake: 4.0

johnwparent opened the pull request:
CMake has released its first major version in a decade!

This major version officially breaks support for multiple features deprecated in 3.5, and will no longer be compatible with projects specifying a required version below 3.5.

This PR:

* Adds CMake 4.0




<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

haampie commented on var/spack/repos/builtin/packages/cmake/package.py:
```diff
@@ -361,6 +362,15 @@ def setup_dependent_package(self, module, dependent_spec):
+        # versions older than 3.5 are actually compatible with newer CMake
+        # and do not use CMake 3.4 or older features. This allows those
+        # projects to use a newer CMake
+        if self.spec.satisfies("@4.0.0-rc1:"):
```
```suggestion
        if self.spec.satisfies("@4:"):
```

johnwparent replied:
Resolved

zackgalbreath commented on var/spack/repos/builtin/packages/cmake/package.py:
```diff
@@ -361,6 +362,15 @@ def setup_dependent_package(self, module, dependent_spec):
+        # versions older than 3.5 are actually compatible with newer CMake
+        # and do not use CMake 3.4 or older features. This allows those
+        # projects to use a newer CMake
+        if self.spec.satisfies("@4.0.0-rc1:"):
```
is this reference to the release candidate intentional, or is it leftover from prior manual testing while we were waiting for CMake 4.0 to be officially released?

johnwparent replied:
Copy and paste error, thanks Zack

johnwparent replied:
Pushed

haampie commented on var/spack/repos/builtin/packages/callpath/package.py:
```diff
@@ -25,6 +25,7 @@ class Callpath(CMakePackage):
     depends_on("adept-utils")
     depends_on("mpi")
     depends_on("cmake@2.8:", type="build")
+    conflicts("cmake@4:")
```
I think more correct is `^cmake@4:`, but as of yesterday `%cmake@4:` should probably be most correct?

In the documentation I put a section on forward and backward compat bounds: https://spack.readthedocs.io/en/latest/packaging_guide.html#specifying-backward-and-forward-compatibility where I recommended `depends_on("cmake@:3", type="build")` for forward compat bounds.

johnwparent replied:
is `%` no longer the compiler interface?
Also this is a direct dependency, not a transitive/provider dep, so are you trying to say for this and anything that depends on it? I don't think that's correct unless it exports some feature that is incompatible w/ CMake 4.0.

alalazo replied:
> but as of yesterday %cmake@4: should probably be most correct?

I'd wait to use this syntax for non-compilers until we get the other parser modifications in. Caveat is that I think this will work for `%cmake@X` but will warn for `%cmake +foo`

haampie replied:
> so are you trying to say for this and anything that depends on it?

No, `conflicts("cmake@4:")` is worse than `conflicts("^cmake@4:")`. The former means it can't be anywhere in the DAG including ancestors, the latter means it can't be anywhere in the sub-dag of direct dependencies and transitive link/run deps.

I would suggest to stick to `depends_on("cmake@:3", when="build_system=cmake", type="build")` for now.

johnwparent replied:
I was actually unaware of the nuances there, thanks for the context. Makes a lot of sense, will fix.

tgamblin replied:
This is no longer part of this PR. Resolving.

johnwparent commented:
@spackbot fix style

johnwparent commented:
@spackbot fix style

luszczek commented:
I've added  the recommended syntax in bed5bc4 to PLASMA's package file to mimic what's already done for MAGMA above.

johnwparent commented:
@spackbot fix style

tgamblin commented:
Does it make sense to split this up, and to put CMake 4.0 in as non-preferred so that people can start trying it out, THEN do all the package modifications?

johnwparent commented:
> Does it make sense to split this up, and to put CMake 4.0 in as non-preferred so that people can start trying it out, THEN do all the package modifications?

So long as the lack of preference for 4.0 strictly keeps it out of the CI pipelines that's probably better. If that's not the case I'd be worried about a spurious 4.0 somehow making it into CI and causing more intermittent failures.

I'll update this PR to prefer CMake 3.3x.x

johnwparent committed:
CMake: 4.0

CMake has released its first major version in a decade!

This major version officially breaks support for multiple features deprecated in 3.5, and will no longer be compatible with projects specifying a required version below 3.5.

This PR:

* Adds CMake 4.0

johnwparent commented:
@tgamblin this PR now _only_ vendors the CMake 4.0 upgrade, but prefers 3.31.6

tgamblin committed:
Merge branch 'develop' into cmake/4.0

johnwparent commented:
@tgamblin this is passing gitlab now, sync script just hasn't picked it up
